File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-tgit/src/main/java/com/tencent/devops/scm/provider/git/tgit/TGitFileService.java
package com.tencent.devops.scm.provider.git.tgit;

import com.tencent.devops.scm.api.FileService;
import com.tencent.devops.scm.api.pojo.Content;
import com.tencent.devops.scm.api.pojo.ContentInput;
import com.tencent.devops.scm.api.pojo.Tree;
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository;
import com.tencent.devops.scm.sdk.tgit.TGitApiFactory;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitRepositoryFile;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitTreeItem;
import java.util.List;
import java.util.stream.Collectors;

public class TGitFileService implements FileService {

    private final TGitApiFactory apiFactory;

    public TGitFileService(TGitApiFactory apiFactory) {
        this.apiFactory = apiFactory;
    }

    @Override
    public Content find(ScmProviderRepository repository, String path, String ref) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitRepositoryFile file = tGitApi.getRepositoryFileApi()
                    .getFile(repo.getProjectIdOrPath(), path, ref);
            return TGitObjectConverter.convertContent(file);
        });
    }

    @Override
    public void create(ScmProviderRepository repository, String path, ContentInput input) {
        TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            tGitApi.getRepositoryFileApi()
                    .createFile(repo.getProjectIdOrPath(), path,
                            input.getRef(), input.getContent(), input.getMessage());
        });
    }

    @Override
    public void update(ScmProviderRepository repository, String path, ContentInput input) {
        TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            tGitApi.getRepositoryFileApi()
                    .updateFile(repo.getProjectIdOrPath(), path,
                            input.getRef(), input.getContent(), input.getMessage());
        });
    }

    @Override
    public List<Tree> listTree(ScmProviderRepository repository, String path, String ref, boolean recursive) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            List<TGitTreeItem> treeItems = tGitApi.getRepositoryFileApi()
                    .getTree(repo.getProjectIdOrPath(), path, ref, recursive);
            return treeItems.stream()
                    .map(TGitObjectConverter::convertTree)
                    .collect(Collectors.toList());
        });
    }
}

File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-tgit/src/main/kotlin/com/tencent/devops/scm/provider/git/tgit/TGitFileService.kt
package com.tencent.devops.scm.provider.git.tgit

import com.tencent.devops.scm.api.FileService
import com.tencent.devops.scm.api.pojo.Content
import com.tencent.devops.scm.api.pojo.ContentInput
import com.tencent.devops.scm.api.pojo.Tree
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository
import com.tencent.devops.scm.sdk.tgit.TGitApiFactory

/**
 * TGit 文件服务实现类
 * @property apiFactory TGit API工厂
 */
class TGitFileService(
    private val apiFactory: TGitApiFactory
) : FileService {

    /**
     * 获取文件内容
     * @param repository 代码仓库信息
     * @param path 文件路径
     * @param ref 分支/标签/提交ID
     * @return 文件内容对象
     */
    override fun find(repository: ScmProviderRepository, path: String, ref: String): Content {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val file = tGitApi.repositoryFileApi.getFile(repo.projectIdOrPath, path, ref)
            TGitObjectConverter.convertContent(file)
        }
    }

    /**
     * 创建文件
     * @param repository 代码仓库信息
     * @param path 文件路径
     * @param input 文件内容输入
     */
    override fun create(repository: ScmProviderRepository, path: String, input: ContentInput) {
        TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            tGitApi.repositoryFileApi.createFile(
                repo.projectIdOrPath,
                path,
                input.ref,
                input.content,
                input.message
            )
        }
    }

    /**
     * 更新文件
     * @param repository 代码仓库信息
     * @param path 文件路径
     * @param input 文件内容输入
     */
    override fun update(repository: ScmProviderRepository, path: String, input: ContentInput) {
        TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            tGitApi.repositoryFileApi.updateFile(
                repo.projectIdOrPath,
                path,
                input.ref,
                input.content,
                input.message
            )
        }
    }

    /**
     * 获取目录树
     * @param repository 代码仓库信息
     * @param path 目录路径
     * @param ref 分支/标签/提交ID
     * @param recursive 是否递归获取
     * @return 目录树列表
     */
    override fun listTree(
        repository: ScmProviderRepository,
        path: String,
        ref: String,
        recursive: Boolean
    ): List<Tree> {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            tGitApi.repositoryFileApi.getTree(repo.projectIdOrPath, path, ref, recursive)
                    .map { TGitObjectConverter.convertTree(it) }
        }
    }
}
